js/driver/ps2/mouse.js

```diff
@@ -0,0 +1,196 @@
+    ySign: split[2],
+    xSign: split[3],
+    alwaysOne: split[4],
+    buttons: {
+      middle: split[5],
+      right: split[6],
+      left: split[7],
+    },
+    xOffset: packet[1],
+    yOffset: packet[2],
+  };
+
+  // just a sanity check:
+  if (info.alwaysOne !== 1) return; // something's wrong with this packet
+
+  if (flags.middle && info.buttons.middle === 0) {
+    flags.middle = false;
+    setImmediate(() => {
+      runtime.mouse.onMouseup.dispatch(1);
+    });
+  } else if (!flags.middle && info.buttons.middle === 1) {
+    flags.middle = true;
+    setImmediate(() => {
+      runtime.mouse.onMousedown.dispatch(1);
+    });
+  } else if (flags.right && info.buttons.right === 0) {
+    flags.right = false;
+    setImmediate(() => {
+      runtime.mouse.onMouseup.dispatch(2);
+    });
+  } else if (!flags.right && info.buttons.right === 1) {
+    flags.right = true;
+    setImmediate(() => {
+      runtime.mouse.onMousedown.dispatch(2);
+    });
+  } else if (flags.left && info.buttons.left === 0) {
+    flags.left = false;
+    setImmediate(() => {
+      runtime.mouse.onMouseup.dispatch(0);
+    });
+  } else if (!flags.left && info.buttons.left === 1) {
+    flags.left = true;
+    setImmediate(() => {
+      runtime.mouse.onMousedown.dispatch(0);
+    });
+  }
+
+  if (info.xOffset !== 0 || info.yOffset !== 0) {
+    setImmediate(() => {
+      runtime.mouse.onMousemove.dispatch({
+        x: (info.xSign) ? (info.xOffset | 0xffffff00) : info.xOffset,
+        y: (info.ySign) ? (info.yOffset | 0xffffff00) : info.yOffset,
+      });
+    });
+  }
+}
+
+const driver = {
+  init(device) {
+    const irq = device.irq;
+    const [mainPort, port64] = device.ioPorts;
+
+    function mouseWait(type) {
+      return new Promise((resolve, reject) => {
+        let maxIter = 1500;
+        function loop() {
+          return new Promise((resolve, reject) => {
+            if (maxIter === 0) return resolve();
+            if (type === 0) {
+              if ((port64.read8() & 1) === 1) {
+                return resolve();
+              }
+            } else {
+              if ((port64.read8() & 2) === 0) {
+                return resolve();
+              }
+            }
+            maxIter--;
+            setImmediate(() => {
+              loop().then(resolve).catch(reject);
+            });
+          });
+        }
+        setImmediate(() => {
+          loop().then(resolve).catch(reject);
+        });
+      });
+    }
+
+    function mouseWrite(data) {
+      return new Promise((resolve, reject) => {
+        mouseWait(1).then(() => {
+          port64.write8(0xd4);
+          return mouseWait(1);
+        }).then(() => {
+          mainPort.write8(data);
+          resolve();
+        }).catch(reject);
+      });
+    }
+
+    function mouseRead() {
+      return new Promise((resolve, reject) => {
+        mouseWait(0).then(() => {
+          resolve(mainPort.read8());
+        }).catch(reject);
+      });
```

Comment:
Ahh, ok. I just tried it out in jsFiddle, I didn't even know that could be done, cool!